### ISSUE TITLE
DeferredContext() should only be created in the context of an eliot.Action.

### ISSUE THREAD

**Issue by jonathanj** [ClusterHQ/eliot#205]
This exception is raised when trying to instantiate `DeferredContext` outside of an action. My current use here is to correctly propagate the Eliot context across long async callback chains. Any Eliot actions created outside of the directly wrapped context will incorrectly become new tasks instead of children of the overarching parent action.

Unfortunately this error means that I *have* to call these functions within an Eliot context. Am I doing this wrong? Should I be creating spurious Eliot actions to appease `DeferredContext`?

**Comment by itamarst** [comment]
Sorry, was on vacation. Will reply when awake enough to comprehend English. :grinning:

**Comment by itamarst** [comment]
Just so I get your use case: you have a function you want to use both within a eliot context and outside an eliot context, but the function uses `DeferredContext` which requires an eliot context?

If I got that right: why do you ever want to call a function outside of an eliot context?

**Comment by jonathanj** [comment]
Yes, that's correct. There are tests for the logging functionality called from another layer up, so the higher level invokes these within an action and there are no test failures. However, after adding logging to existing code I found the existing low-level tests were all breaking. 

It's not that I want to, it's that it seems unnecessary. For example, the code in this case is a high-level wrapper, the methods of which are quite simple:

```python
def get_foos(self):
  return self._makeRequestAndProcessTheResult('GET', '/foos')
```

`_makeRequestAndProcessTheResult` logs the request with enough information to understand how and why it was called. Having to wrap each function in a superfluous action is just not terribly pleasant to look at:

```python
def get_foos(self):
  action = LOG_ACTION(...)
  with action.context():
    d = DeferredContext(self._makeRequestAndProcessTheResult('GET', '/foos')
    d.addActionFinish()
    return d.result
```

Perhaps one could argue for making sure `DeferredContext` is only ever called in the context of an existing action, even if that means creating a superfluous one to guarantee that. I found this behaviour somewhat surprising though.

**Comment by itamarst** [comment]
The basic use case DeferredContext was written for is "I have an action I just started, I want to run these callbacks on a Deferred within action context and then finish the action when the Deferred fires." So typically you'd start an action just where you did the DeferredContext.

In this case... it sounds like you're not starting the action in that method? And moreover not calling `addActionFinish`? If you are calling `addActionFinish` that's a bit of a weird thing to do in a function that didn't start an action.

You showed the code without logging and the code with superflous-action logging, could you show me the code as you wanted to write it?

**Comment by jonathanj** [comment]
Indeed I do not call `addActionFinish` because like you say that's a not something a function that didn't start an action should do.

The code as it was before I realised it was broken (which is probably how I would have wanted to write it) was:

```python
def get_foos(self):
  return DeferredContext(self._request('GET', '/foos')).result
```

Which I'll grant you is kind of weird code but is something you may want to do in a complex callback chain too:

```python
def _attach_bars(foos):
  # Pretend this is a complex callback that needs "foos" as some kind of internal state such that we can't simply chain things to the original DeferredContext.
  d = DeferredContext(self.get_bars())
  d.addCallback(zip, foos) 
  # No addActionFinish() because we didn't create our own action since we're just an enabler.
  return d.result

def bars_and_foos(self):
  return self.get_foos().addCallback(_attach_bars)
```

Possibly the correct thing to do here is make sure that these methods always create an action whenever you create a `DeferredContext` but that can easily end up with a large number of actions that provide no real clarity.

**Comment by itamarst** [comment]
OK, I think I get the use case. I guess there's no particular reason why `DeferredContext` shouldn't effectively be a no-op if there's no action in context.

**Comment by tomprince** [comment]
Note that
```
DeferredContext(self._request('GET', '/foos')).result
```
is a no-op (other than the validation). `DeferredContext` can't reach into the the call to `self._request` and make any callbacks added there get run in the current action.
